File: flutter-daro-sdk-a/android/src/main/kotlin/com/example/flutter_daro_sdk/FlutterDaroSdkPlugin.kt
package com.example.flutter_daro_sdk

import android.app.Activity
import droom.daro.Daro
import droom.daro.core.model.DaroAdDisplayFailError
import droom.daro.core.model.DaroAdInfo
import droom.daro.core.model.DaroAdLoadError
import droom.daro.core.model.DaroRewardedAd
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutterDaroSdkPlugin */
class FlutterDaroSdkPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  private lateinit var methodChannel : MethodChannel

  /// The EventChannel for sending events to Flutter
  private lateinit var eventChannel : EventChannel
  private var eventSink: EventChannel.EventSink? = null
  private lateinit var bannerAdFactory: FlutterDaroBannerAdFactory

  /// Current activity reference
  private var activity: Activity? = null

  /// 앱 카테고리 타입
  private var appCategory: String? = null

  /// 리워드 광고 인스턴스 맵
  private val rewardAdMap = mutableMapOf<String, FlutterDaroRewardAd>()

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.daro.flutter_daro_sdk/channel")
    methodChannel.setMethodCallHandler(this)

    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.daro.flutter_daro_sdk/events")
    eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
      override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
      }

      override fun onCancel(arguments: Any?) {
        eventSink = null
      }
    })

    bannerAdFactory = FlutterDaroBannerAdFactory.registerWith(flutterPluginBinding)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "initialize" -> {
        initialize(call, result)
      }
      "loadRewardAd" -> {
        loadRewardAd(call, result)
      }
      "showRewardAd" -> {
        showRewardAd(call, result)
      }
      "disposeRewardAd" -> {
        disposeRewardAd(call, result)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  /// SDK 초기화
  private fun initialize(call: MethodCall, result: Result) {
    try {
      val args = call.arguments as? Map<*, *> ?: return result.error(
        "1001",
        "INVALID_ARGUMENT",
        "Invalid arguments for initialize",
      )

      appCategory = args["appCategory"] as? String
      val currentActivity = activity ?: return result.success(false)

      // 문서 참고: https://guide.daro.so/ko/sdk-integration/android/get-started#sdk-%EC%B4%88%EA%B8%B0%ED%99%94%ED%95%98%EA%B8%B0
      val configBuilder = Daro.SDKConfig.Builder()
      val options = args["options"] as? Map<*, *>
      if (options != null) {
        when (options["logLevel"] as? String) {
          "off" -> configBuilder.setDebugMode(false)
          "error" -> configBuilder.setDebugMode(true)
          "debug" -> configBuilder.setDebugMode(true)
          else -> {}
        }

        when(options["appMute"] as? Boolean) {
          true -> configBuilder.setAppMute(true)
          false -> configBuilder.setAppMute(false)
          else -> {}
        }
      }

       Daro.init(
         application = currentActivity.application,
         sdkConfig = configBuilder.build()
       )

      // 초기화 성공
      result.success(true)
    } catch (e: Exception) {
      result.error("1002", "INIT_ERROR", e.message)
    }
  }

  private fun createAdInstance(
    context: Activity,
    adType: FlutterDaroRewardAdType,
    adUnit: String,
    placement: String?,
    options: Map<*,*>?
  ): FlutterDaroRewardAd {
    rewardAdMap[adUnit]?.let {
      return it
    }
    val result = FlutterDaroRewardAdFactory.create(
      context = context,
      adType = adType,
      adUnit = adUnit,
      placement = placement,
      options = options,
      loadListener = object: FlutterDaroRewardAdLoadListener {
        override fun onAdLoadSuccess(ad: FlutterDaroRewardAd, adInstance: Any?, adInfo: DaroAdInfo?) {
          sendRewardAdEvent(
            adUnit,
            "onAdLoadSuccess",
            mapOf(
              "latency" to (adInfo)?.latency,
              "ad" to adInstance?.toString()
            )
          )
        }
        override fun onAdLoadFail(error: DaroAdLoadError) {
          sendRewardAdEvent(
            adUnit,
            "onAdLoadFail",
            mapOf(
              "code" to error.code,
              "message" to error.message
            )
          )
        }
      }
    )
    rewardAdMap[adUnit] = result
    return result
  }

  /// 리워드 광고 로드
  private fun loadRewardAd(call: MethodCall, result: Result) {
    try {
      val args = call.arguments as? Map<*, *> ?: return result.error(
        "1001",
        "INVALID_ARGUMENT",
        "Invalid arguments for loadRewardAd",
      )

      val adType = FlutterDaroRewardAdType.fromString(args["adType"] as? String?) ?: return result.error(
        "1001",
        "INVALID_ARGUMENT",
        "adType is required ${args["adType"]}",
      )

      val adUnit = args["adUnit"] as? String ?: return result.error(
        "1001",
        "INVALID_ARGUMENT",
        "adUnit is required",
      )

      val placement = args["placement"] as? String

      val currentActivity = activity
        ?: return result.error(
          "1013",
          "NO_ACTIVITY",
          "No activity available to load ad",
        )

      val options = args["options"] as? Map<*, *>


      // 새로운 리워드 광고 인스턴스 생성
      val adInstance = createAdInstance(currentActivity, adType, adUnit, placement, options)

      // 광고 로드
      adInstance.loadAd(
        autoShow = false,
        listener = null,
        result = { success, data ->
          result.success(success)
        }
      )
    } catch (e: Exception) {
      result.success(false)
    }
  }

  /// 리워드 광고 표시
  private fun showRewardAd(call: MethodCall, result: Result) {
    try {
      val currentActivity = activity
        ?: return result.error(
          "1013",
          "NO_ACTIVITY",
          "No activity available to show ad",
        )

      val args = call.arguments as? Map<*, *> ?: return result.error(
        "1001",
        "INVALID_ARGUMENT",
        "Invalid arguments for showRewardAd",
      )

      val adType = FlutterDaroRewardAdType.fromString(args["adType"] as? String?) ?: return result.error(
        "1001",
        "INVALID_ARGUMENT",
        "adType is required : ${args["adType"]}",
      )
      val adUnit = args["adUnit"] as? String
        ?: return result.error(
          "1001",
          "INVALID_ARGUMENT",
          "adUnit is required",
        )
      val placement = args["placement"] as? String
      val options = args["options"] as? Map<*, *>

      var adInstance = rewardAdMap[adUnit]
      if (adInstance == null) {
        adInstance = createAdInstance(currentActivity, adType, adUnit, placement, options)
      }

      adInstance.showAd(
        listener = object : FlutterDaroRewardAdListener {
          override fun onShown(adInfo: DaroAdInfo) {
            sendRewardAdEvent(adUnit, "onShown", emptyMap())
          }
          override fun onRewarded(adInfo: DaroAdInfo, rewardItem: DaroRewardedAd.DaroRewardedItem) {
            sendRewardAdEvent(
              adUnit,
              "onRewarded",
              mapOf(
                "type" to rewardItem.type,
                "amount" to rewardItem.amount
              )
            )
          }
          override fun onDismiss(adInfo: DaroAdInfo) {
            sendRewardAdEvent(adUnit, "onDismiss", emptyMap())
            dispose(adUnit)
          }
          override fun onFailedToShow(adInfo: DaroAdInfo, error: DaroAdDisplayFailError) {
            sendRewardAdEvent(
              adUnit,
              "onFailedToShow",
              mapOf(
                "code" to -1,
                "message" to error.toString(),
              ),
            )
            dispose(adUnit)
          }
          override fun onAdImpression(adInfo: DaroAdInfo) {
            sendRewardAdEvent(adUnit, "onAdImpression", emptyMap())
          }
          override fun onAdClicked(adInfo: DaroAdInfo) {
            sendRewardAdEvent(adUnit, "onAdClicked", emptyMap())
          }
        },
        result = {success, data ->
          result.success(success)
        }
      )
    } catch (e: Exception) {
      result.success(false)
    }
  }

  /// 리워드 광고 이벤트를 Flutter로 전송
  private fun sendRewardAdEvent(adUnit: String, eventType: String, data: Map<String, Any?>) {
    eventSink?.success(mapOf(
      "eventName" to eventType,
      "adUnit" to adUnit,
      "data" to data,
    ))
  }

  private fun dispose(adUnit: String) {
    rewardAdMap.remove(adUnit)?.destroy()
  }
  /// 리워드 광고 인스턴스 해제
  private fun disposeRewardAd(call: MethodCall, result: Result) {
    try {
      val args = call.arguments as? Map<*, *> ?: return result.error(
        "1001",
        "INVALID_ARGUMENT",
        "Invalid arguments for disposeRewardAd",
      )

      val adUnit = args["adUnit"] as? String ?: return result.error(
        "1001",
        "INVALID_ARGUMENT",
        "adUnit is required",
      )

      dispose(adUnit)

      result.success(null)
    } catch (e: Exception) {
      result.error("1030", "DISPOSE_ERROR", e)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
    bannerAdFactory.onDestroy()
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
    // 모든 리워드 광고 인스턴스 해제
    rewardAdMap.values.forEach { it.destroy() }
    rewardAdMap.clear()
  }
}

File: flutter-daro-sdk-a/ios/Classes/FlutterDaroSdkPlugin.swift
import Flutter
import UIKit
import Daro

public class FlutterDaroSdkPlugin: NSObject, FlutterPlugin {
  /// MethodChannel for method calls from Flutter
  private var methodChannel: FlutterMethodChannel?
  
  /// EventChannel for sending events to Flutter
  private var eventChannel: FlutterEventChannel?
  private var eventSink: FlutterEventSink?
  
  /// 리워드 광고 인스턴스 맵 (타입과 adUnit를 조합한 키 사용: "type:adUnit")
  private var rewardAdMap: [String: FlutterDaroRewardAd] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = FlutterDaroSdkPlugin()
    
    // MethodChannel 설정
    let methodChannel = FlutterMethodChannel(
      name: "com.daro.flutter_daro_sdk/channel",
      binaryMessenger: registrar.messenger()
    )
    instance.methodChannel = methodChannel
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    
    // EventChannel 설정
    let eventChannel = FlutterEventChannel(
      name: "com.daro.flutter_daro_sdk/events",
      binaryMessenger: registrar.messenger()
    )
    instance.eventChannel = eventChannel
    eventChannel.setStreamHandler(instance)
    FlutterDaroBannerFactory.register(with: registrar)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      initialize(call: call, result: result)
    case "setOptions":
      setOptions(call: call, result: result)
    case "loadRewardAd":
      loadRewardAd(call: call, result: result)
    case "showRewardAd":
      showRewardAd(call: call, result: result)
    case "disposeRewardAd":
      disposeRewardAd(call: call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// SDK 초기화
  private func initialize(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(
        code: "INVALID_ARGUMENT",
        message: "Invalid arguments for initialize",
        details: nil
      ))
      return
    }
    
    // appCategory = args["appCategory"] as? String
    // 문서 참고: https://guide.daro.so/ko/sdk-integration/ios_new/get-started#sdk-%EC%B4%88%EA%B8%B0%ED%99%94%ED%95%98%EA%B8%B0    
    DaroAds.shared.initialized { error in
      if let error {
        result(FlutterError(
          code: "INITIALIZE_FAILED",
          message: "Daro SDK initilized error : \(error)",
          details: nil
        ))
      } else {
        let options = args["options"] as? [String: Any]
        if let options {
          self._setOptions(
            userId: options["userId"] as? String, 
            logLevel: options["logLevel"] as? String, 
            appMute: options["appMute"] as? Bool
          )
        }
        result(true)
      }
    }
  }

  /// SDK 옵션 설정
  private func setOptions(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(
        code: "INVALID_ARGUMENT",
        message: "Invalid arguments for setOptions",
        details: nil
      ))
      return
    }
    
    self._setOptions(
      userId: args["userId"] as? String, 
      logLevel: args["logLevel"] as? String, 
      appMute: args["appMute"] as? Bool
    )

    result(true)
  }

  private func _setOptions(userId: String?, logLevel: String?, appMute: Bool?) {
    if let userId {
      DaroAds.shared.userId = userId
    }
    if let logLevel {
      switch logLevel {
        case "off":
          DaroAds.shared.logLevel = .off
        case "error":
          DaroAds.shared.logLevel = .error
        case "debug":
          DaroAds.shared.logLevel = .debug
        default:
          print("Invalid logLevel: \(logLevel)")
      }
    }
    if let appMute {
      DaroAds.shared.setAppMuted(appMute)
    }
  }
  
  /// 리워드 광고 이벤트를 Flutter로 전송
  private func sendRewardAdEvent(adUnit: String, eventType: String, data: [String: Any?] = [:]) {
    let event: [String: Any] = [
      "eventName": eventType,
      "adUnit": adUnit,
      "data": data
    ]
    eventSink?(event)
  }

  /// 리워드 광고 인스턴스 생성
  private func _createRewardAdInstance(
    adType: FlutterDaroRewardAdType, 
    adUnit: String, 
    placement: String?, 
    options: [String: Any]? = nil
  ) -> FlutterDaroRewardAd {
    if let adInstance = rewardAdMap[adUnit] {
      return adInstance
    }
    let adInstance = FlutterDaroRewardAdFactory.create(
      adType: adType,
      adUnit: adUnit,
      placement: placement,
      options: options,
      listener: FlutterDaroRewardAdLoadListener(
        onAdLoadSuccess: { adItem, ad, adInfo in
          self.sendRewardAdEvent(adUnit: adUnit, eventType: "onAdLoadSuccess", data: [
            "ad": String(describing: ad as Any),
            "adInfo": String(describing: adInfo as Any)
          ])
        },
        onAdLoadFail: { error in
          self.sendRewardAdEvent(adUnit: adUnit, eventType: "onAdLoadFail", data: [
            "error": error.localizedDescription
          ])
          self._dispose(adUnit: adUnit)
        },
        onAdImpression: { adInfo in
          self.sendRewardAdEvent(adUnit: adUnit, eventType: "onAdImpression")
        },
        onAdClicked: { adInfo in
          self.sendRewardAdEvent(adUnit: adUnit, eventType: "onAdClicked")
        }
      )
    )
    self.rewardAdMap[adUnit] = adInstance
    return adInstance
  }
  
  /// 리워드 광고 로드
  private func loadRewardAd(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(
        code: "INVALID_ARGUMENT",
        message: "Invalid arguments for loadRewardAd",
        details: nil
      ))
      return
    }
    
    guard let adType = args["adType"] as? String else {
      result(FlutterError(
        code: "INVALID_ARGUMENT",
        message: "adType is required",
        details: nil
      ))
      return
    }
    
    guard let adUnit = args["adUnit"] as? String else {
      result(FlutterError(
        code: "INVALID_ARGUMENT",
        message: "adUnit is required",
        details: nil
      ))
      return
    }
    
    let placement = args["placement"] as? String
    let options = args["options"] as? [String: Any]
    
    // String을 enum으로 변환
    guard let rewardAdType = FlutterDaroRewardAdType(rawValue: adType) else {
      result(FlutterError(
        code: "INVALID_ARGUMENT",
        message: "Invalid adType: \(adType)",
        details: nil
      ))
      return
    }
    
    // 새로운 리워드 광고 인스턴스 생성
    let adInstance: FlutterDaroRewardAd = _createRewardAdInstance(adType: rewardAdType, adUnit: adUnit, placement: placement, options: options)
    
    // 광고 로드
    adInstance.loadAd() { success, error in
      if success {
        result(nil)
      } else {
        result(FlutterError(
          code: "LOAD_ERROR",
          message: error?.localizedDescription ?? "Failed to load ad",
          details: nil
        ))
      }
    }
  }
  
  /// 리워드 광고 인스턴스 표시
  private func showRewardAd(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(false)
      return
    }
    
    guard let adType = args["adType"] as? String else {
      result(false)
      return
    }
    
    guard let adUnit = args["adUnit"] as? String else {
      result(false)
      return
    }
    
    let placement = args["placement"] as? String
    
    // String을 enum으로 변환
    guard let rewardAdType = FlutterDaroRewardAdType(rawValue: adType) else {
      result(false)
      return
    }
    
    var adInstance = rewardAdMap[adUnit]
    
    // 인스턴스가 없으면 자동으로 생성하고 로드
    if adInstance == nil {
      // 새로운 리워드 광고 인스턴스 생성
      let options = args["options"] as? [String: Any]
      adInstance = _createRewardAdInstance(adType: rewardAdType, adUnit: adUnit, placement: placement, options: options)
    }
    
    // 인스턴스가 있으면 바로 표시
    adInstance?.showAd(
      listener: FlutterDaroRewardAdListener(
        onShown: { adInfo in
          self.sendRewardAdEvent(adUnit: adUnit, eventType: "onShown")
        },
        onRewarded: { adInfo, reward in
          self.sendRewardAdEvent(adUnit: adUnit, eventType: "onRewarded", data: [
            "reward": [
              "amount": reward?.amount ?? 0,
              "type": reward?.rewardType ?? ""
            ]
          ])
        },
        onDismiss: { adInfo in
          self.sendRewardAdEvent(adUnit: adUnit, eventType: "onDismiss")
          self._dispose(adUnit: adUnit)
        },
        onFailedToShow: { adInfo, error in
          self.sendRewardAdEvent(
            adUnit: adUnit, 
            eventType: "onFailedToShow", 
            data: ["error": error.localizedDescription]
          )
          self._dispose(adUnit: adUnit)
        }
    )) { success, error in
      result(success ? true : FlutterError(
        code: "SHOW_ERROR",
        message: "Failed to show ad: \(error)",
        details: nil
      ))
    }
  }
  
  /// 리워드 광고 인스턴스 해제
  private func _dispose(adUnit: String) {
    rewardAdMap[adUnit]?.destroy()
    rewardAdMap.removeValue(forKey: adUnit)
  }
  private func disposeRewardAd(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(
        code: "INVALID_ARGUMENT",
        message: "Invalid arguments for disposeRewardAd",
        details: nil
      ))
      return
    }
    
    guard let adUnit = args["adUnit"] as? String else {
      result(FlutterError(
        code: "INVALID_ARGUMENT",
        message: "adUnit is required",
        details: nil
      ))
      return
    }
    
    self._dispose(adUnit: adUnit)
    
    result(nil)
  }
}

// MARK: - FlutterStreamHandler
extension FlutterDaroSdkPlugin: FlutterStreamHandler {
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    self.eventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    // 모든 리워드 광고 인스턴스 해제
    rewardAdMap.values.forEach { $0.destroy() }
    rewardAdMap.removeAll()
    return nil
  }
}
